elastic/obs-infraobs-integrations: Address skipped validation SVR00002

shmsr opened the pull request:
I'll break down the relevant parts of the filter added in the dashboards (a filter looks like this when exported):

```json
"filter": [
    {
        "$state": {
            "store": "appState"
        },
        "meta": {
            "alias": null,
            "disabled": false,
            "indexRefName": "kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index",
            "key": "data_stream.dataset",
            "negate": false,
            "params": {
                "query": "airflow.statsd"
            },
            "type": "phrase"
        },
        "query": {
            "match_phrase": {
                "data_stream.dataset": "airflow.statsd"
            }
        }
    }
],
```

- "filter": An array containing filter configurations used to narrow down data in visualizations or dashboards.

    - "meta": Metadata about the filter.
        - "indexRefName": A reference to an index pattern. It's set to "kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index" and points to the index pattern associated with the filter. If there are already a few filters, depends on the position [0] can change. But as SVR00002 means there's no filter there in the filter configuration, it would always be [0] in our case.
        - "key": The field on which the filter operates, specified as "data_stream.dataset".
        - "params": Contains filter-specific parameters. It includes a "query" parameter with a value of "airflow.statsd", indicating that the filter matches data where the "data_stream.dataset" field equals "airflow.statsd".
        - "type": Specifies the filter type, set to "phrase", indicating it's a phrase filter. ([ref](https://www.elastic.co/guide/en/beats/packetbeat/current/kibana-queries-filters.html#_string_queries))

    - "query": Defines the filter's query, which is a "match_phrase" query matching the "data_stream.dataset" field with the value "airflow.statsd".

This filter is used to filter data within a Kibana dashboard based on the "data_stream.dataset" field and the phrase "airflow.statsd". The "indexRefName" references the index pattern associated with the filter, allowing you to specify the index-pattern within the dashboard where the filter applies.


Reference added for `kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index`
```json
{
            "id": "metrics-*",
            "name": "kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index",
            "type": "index-pattern"
},
```

- "id": This specifies the ID of the reference. Contains the index-pattern.

- "name": This specifies the name of the reference. This name is a reference to the index pattern associated with the filter in the dashboard's search source configuration defined earlier.

- "type": This indicates the type of the reference. Reference is associated with an index pattern i.e., metric-*


You can associate an index pattern with the filter in the dashboard and enable the dashboard to use data from that specific index pattern when rendering visualizations.

How each filter is added per dashboard:

- Check the dataset types (logs or metrics) and based on that decide the index-pattern (also other references in the dashboard gives a clue)
- Check the fields used in the dashboard (they belong to which dataset)
- If per dashboard, only a single dataset is used then add a global filter for that dataset and add the reference.

Note: All filters were added manually. I also checked with older stack versions 8.0.0, 8.3.0, etc. to see how the filters are being added. Didn't use elastic-package's export dashboards as it changes a lot more things which would make it very hard to review. Out of all packages that have SVR00002, the lowest kibana.version in the manifest is 8.0.0 and that's why patterns of the filters are checked only until 8.0.0.

Please also see https://github.com/elastic/integrations/pull/8348#issuecomment-1790196842  before reviewing as I have written some rough notes which might be helpful.

## Proposed commit message

Add a global filter to the dashboard and remove the subfilters on the dataset and a global filter on the dataset is introduced. Filters on data_stream.dataset are applied with a phrase type filter on the dataset that is being used in that particular dashboard.


## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] Do a sampling and pick packages. Test them by adding some data to see if all viz are working as expected. Need help from reviewers on this to distribute work.

## Related issues

Relates https://github.com/elastic/integrations/issues/8028

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8348/detail/PR-8348/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8348/detail/PR-8348/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8348/detail/PR-8348/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8348/detail/PR-8348/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_8348.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-10-31T11:27:05.405Z&rangeTo=2023-10-31T11:47:05.405Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=f4334eb119c67e43ed6ac69f091bc87e&transactionId=af54759c6c925eba)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-10-31T11:37:05.405+0000


* Duration: 15 min 2 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2  |
| Skipped      | 0 |
| Total        | 2   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`0/0`) | :green_heart: 
Files | 100.0% (`0/0`) | :green_heart: 
Classes | 100.0% (`0/0`) | :green_heart: 
Methods | 25.0% (`1/4`) | :grey_exclamation: 
Lines | 100.0% (`0/0`) | :green_heart: 
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

tommyers-elastic commented on packages/azure_functions/changelog.yml:
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "0.2.1"
+  changes:
+    - description: Address SVR00002 validation skipped during package-spec v3 migration
```
please can we make the changelog more user-friendly? something like "add global dashboard filter for dataset to improve performance (allows removal of 'SVR00002' validation exception)"

shmsr replied:
Ah, cool. I'll do that.

tommyers-elastic commented on packages/cassandra/kibana/dashboard/cassandra-25b7d6d0-1c71-11ec-84f1-e1733c643874.json:
```diff
@@ -3338,6 +3008,11 @@
         "dashboard": "8.3.0"
     },
     "references": [
+        {
```
we can also remove the references which are no longer used after removal of panel-level filter. i think the 'dangling reference' check would flag these up anyway.

shmsr replied:
I also thought that. But no dangling references errors are shown by elastic-package lint. So seemed risky to do it manually.

shmsr replied:
Ah, I see. Check this: https://github.com/elastic/package-spec/blob/48461578910c405fcc6a92b97a1e558aeb8fb860/code/go/internal/validator/semantic/validate_kibana_no_dangling_object_ids.go#L24

This is why dangling references when type is index-pattern, is ignored.

shmsr replied:
I'll see if I can add the filtering logic (new logic for index-pattern needs to be written) to the program I wrote for https://github.com/elastic/integrations/pull/8347

tommyers-elastic reviewed (commented):
👍 we also need to remove the SVR0002 check from validations.yml

shmsr commented:
> 👍 we also need to remove the SVR0002 check from validations.yml

Yeah, I'll run a script at the end that would run lint again and update all validation.yml. Will do that in the end; right before marking this PR ready for review.

shmsr commented:
Rough notes. Please read them once before reviewing.

- [ ] apache_spark 
    - sub-filter
- [ ] citrix_adc
    - sub-filter with different dataset in 1 dashboard
- [ ] ceph 
    - sub-filter with different dataset in 1 dashboard
- [x] coredns 
    - left some viz filter as they are not exactly the same and using minimum_should_match, etc.
- [ ] couchbase 
    - sub-filter with different dataset in 1 dashboard
- [ ] etcd 
    - not sure what to filter with; multiple dataset
- [x] hadoop 
    - test with 8.10.2 stack once
- [x] iis (please test Dashboard for the Logs IIS integration)
    - this dashboard shows process and cache metrics for the IIS server.
    - this dashboard shows metrics for the IIS server.
    - can the description be better?
- [x] golang
    - test golang
- [ ] kafka
    - only 1 out of 2 dashboard taken care of
- [ ] memcached
    - noticed this in the dashboard:
        - part of Bytes Read (Kb)
        - unit is Kilobits and written bytes instead. Confusing? More such examples are there.
- [ ] microsoft_sqlserver
    - take transaction_log for example. Go to fields/fields.yml
    - fields are like this mssql.metrics.<field_name> whereas I expected mssql.transaction_log.<field_name>
- [ ] mongodb (multiple datasets + subfilter)
- [ ] mysql (multiple datasets)
- [ ] nginx (multiple datasets)
- [ ] oracle (multiple datasets)
- [ ] oracle_weblogic (multiple datasets)
- [ ] postgresql
    - found a custom logic to filter with prefix (not a problem; just taking notes)
    - multiple datasets
- [ ] rabbitmq
    - filter by multiple datasets
- [ ] redis
    - filter by multiple datasets
- [ ] redisenterprise
    - multiple datasets
- [ ] salesforce
    - multiple dataset
- [ ] spring_boot
    - access logs dashboard; i don’t see a dataset. is it correct?
- [ ] vsphere
    - multiple dataset

shmsr commented:
I'll take care of the other packages in another PR.

lalit-satapathy commented:
CC: @harnish-elastic @aliabbas-elastic to review the dashboard changes as part of the filters being added.

harnish-elastic commented:
Why the screenshots/images of the dashboard is not updated in all packages? The look of the dashboard changes if we drag the query to the dashboard level!

shmsr commented:
> Why the screenshots/images of the dashboard are not updated in all packages? The look of the dashboard changes if we drag the query to the dashboard level!

@harnish-elastic Good point. These changes are not done by opening the dashboard and editing from there and then saving from there. Changes are done by hand directly (manually checking *.json, editing, and elastic-package install to check if there are no obvious errors)  in *.json files. As the fix had to be applied to ~35 packages which is a lot; not all could be addressed though. The screenshot was only possible for every package if I had done it via the normal way which would have taken a lot of time. So, during testing, we're loading the dashboard, can we click the screenshot?

aliabbas-elastic reviewed (changes requested):
Requesting for changes according to this [comment](https://github.com/elastic/obs-infraobs-team/issues/1263#:~:text=Kibana%20version%208.11.0-,Issues%20found,-1.%20Cassandra)

mrodm commented:
Hi @shmsr, please update your branch with the latest contents from main branch. There was an important PR merged updating the CI pipelines. Thanks!

shmsr commented:
I have resolved the merge conflicts and also addressed the issues pointed out by @aliabbas-elastic in https://github.com/elastic/obs-infraobs-team/issues/1263#:~:text=Kibana%20version%208.11.0-,Issues%20found,-1.%20Cassandra

aliabbas-elastic reviewed (approved):
LGTM

harnish-elastic reviewed (approved):
LGTM! Still concerned about how we can proceed to update the dashboard screenshot as per this [comment](https://github.com/elastic/integrations/pull/8348#issuecomment-1835606920)!

shmsr commented:
Looks like there are very complex merge conflicts because of recent changes in listed packages. Have to rework on them from scratch.

Here's a list that needs rework from scratch:

- azure_functions
- hadoop
- iis
- traefik
- websphere_application_server

ishleenk17 commented:
@shmsr : Is this still valid ?
